### DISCUSSION TITLE
Terraform pessimistic constraint operator not respected

### DISCUSSION THREAD

**Discussion by ryancausey** [renovatebot/renovate#11701] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:27.4.0-slim@sha256:3d9db475f006f9ff166464797ed5d34b0b24a2a9b12b228f166bbde99546bc6c

### Describe the bug

When using the [pessimistic constraint operator](https://www.terraform.io/docs/language/expressions/version-constraints.html#version-constraint-syntax) with only two digits, I.E. `~> 3.0`, renovate does not pick up on new provider versions and properly update them as `terraform init -update` does.

Here's the sample repository: https://gitlab.com/rccausey/terraform-pessimistic-version-constraint
Here's the MR opened by renovate: https://gitlab.com/rccausey/terraform-pessimistic-version-constraint/-/merge_requests/1

In this case, the locked registry.terraform.io/hashicorp/aws provider version is `3.46.0` and the constraint is `~> 3.0`.

### Expected behavior:

The MR is opened with a change to update to the latest registry.terraform.io/hashicorp/aws provider version 3.58.0 (current as of this writing).

### Actual behavior:

The MR is opened without any updates detected as being needed.

Not sure if this is still relevant but on the renovate job I have `RENOVATE_X_TERRAFORM_LOCK_FILE` set to `true`.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Found 1 package file(s) (repository=rccausey/terraform-pessimistic-version-constraint)
 INFO: Dependency extraction complete (repository=rccausey/terraform-pessimistic-version-constraint)
       "baseBranch": "master",
       "stats": {
         "managers": {"terraform": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
DEBUG: terraform-provider.getDependencies() (repository=rccausey/terraform-pessimistic-version-constraint)
       "lookupName": "hashicorp/aws"
DEBUG: Package releases lookups complete (repository=rccausey/terraform-pessimistic-version-constraint)
       "baseBranch": "master"
DEBUG: branchifyUpgrades (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: 0 flattened updates found:  (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Returning 0 branch(es) (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: config.repoIsOnboarded=false (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: packageFiles with updates (repository=rccausey/terraform-pessimistic-version-constraint)
       "config": {
         "terraform": [
           {
             "packageFile": "main.tf",
             "deps": [
               {
                 "currentValue": "~> 3.0",
                 "depType": "required_provider",
                 "depName": "aws",
                 "datasource": "terraform-provider",
                 "lookupName": "hashicorp/aws",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "hashicorp",
                 "sourceUrl": "https://github.com/hashicorp/terraform-provider-aws",
                 "homepage": "https://registry.terraform.io/providers/hashicorp/aws",
                 "currentVersion": "3.58.0"
               },
               {
                 "depType": "provider",
                 "depName": "aws",
                 "datasource": "terraform-provider",
                 "depIndex": 1,
                 "updates": [],
                 "warnings": [],
                 "versioning": "hashicorp",
                 "skipReason": "invalid-value"
               }
             ]
           }
         ]
       }
DEBUG: ensureOnboardingPr() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: getBranchPr(renovate/configure) (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: findPr(renovate/configure, undefined, open) (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Filling in onboarding PR template (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: getConfigDesc() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: getDescriptionArray() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: getScheduleDesc() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: No schedule (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Found description array (repository=rccausey/terraform-pessimistic-version-constraint)
       "length": 12
DEBUG: getPrList() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Creating onboarding PR (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Creating Merge Request: Configure Renovate (repository=rccausey/terraform-pessimistic-version-constraint)
 INFO: Onboarding PR created (repository=rccausey/terraform-pessimistic-version-constraint)
       "pr": "Merge Request #1"
DEBUG: processRepo() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Repo is onboarding - skipping dependency dashboard (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Removing any stale branches (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: config.repoIsOnboarded=false (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Branch lists (repository=rccausey/terraform-pessimistic-version-constraint)
       "branchList": ["renovate/configure"],
       "renovateBranches": ["renovate/configure"]
DEBUG: remainingBranches= (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: No branches to clean up (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: ensureIssueClosing() (repository=rccausey/terraform-pessimistic-version-constraint)
DEBUG: Repository timing splits (milliseconds) (repository=rccausey/terraform-pessimistic-version-constraint)
       "splits": {"init": 5379, "extract": 427, "lookup": 328, "update": 822},
       "total": 6968
DEBUG: http statistics (repository=rccausey/terraform-pessimistic-version-constraint)
       "hostStats": {
         "api.github.com": {"requestCount": 1, "requestAvgMs": 265, "queueAvgMs": 2},
         "gitlab.com": {"requestCount": 4, "requestAvgMs": 374, "queueAvgMs": 2},
         "registry.terraform.io": {
           "requestCount": 2,
           "requestAvgMs": 108,
           "queueAvgMs": 1
         }
       },
       "totalRequests": 7
 INFO: Repository finished (repository=rccausey/terraform-pessimistic-version-constraint)
       "durationMs": 6968
DEBUG: Renovate exiting
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by viceice** [comment]
https://docs.renovatebot.com/configuration-options/#rangestrategy

**Reply by ryancausey** [reply]
@viceice I've tested all the different settings for that value, and the only one that caused renovate to consider dependencies changed was the `pin` option. That's not what I want, and renovate not finding dependency updates on the other flag values seems like incorrect behavior to me. I would've expected the terraform manager to automatically follow the version constraints in the required providers block as well as the .terraform.lock.hcl file, especially when on the `auto` setting.

Am I doing something wrong? This is the renovate.json contents:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "rangeStrategy": "auto"
}
```

**Reply by rarkins** [reply]
I'm not certain that I have understood everything, but it sounds like you may be expecting `rangeStrategy=update-lockfile` behavior and it's not yet supported. @secustor wdyt?

**Reply by secustor** [reply]
I had to look into the range strategy `update-lockfile` as I have not used it so far. Renovate seems to expect that `lockedVersion` is set during dependency extraction, which is not done right now. Therefore it is not supported right now.    

@ryancausey  as alternative you can use [lockfileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) so far

**Reply by ryancausey** [reply]
@rarkins I'm not necessarily looking to use a specific range strategy. I just want to use one that lets me have a pessimistic version specifier, `~> 3.0` in this case, work as expected. If there's a specific range strategy that achieves that I'll use it, but none of them have seemed to do that so far. The bug I filed originally should describe the behavior of the pessimistic version specifier as it behaves when doing a `terraform init -upgrade`. I expected the terraform manager for semantic-release to mirror the update behavior of terraform by default, which it doesn't appear to do.

@secustor I will try that setting out and report back.

**Reply by viceice** [reply]
@ryancausey `lockFileMaintenance` will do it. It updates all locked versions to latest in-range version.

**Reply by ryancausey** [reply]
@rarkins @viceice @secustor thanks for the help! Using the below renovate.json made it work for me.

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "lockFileMaintenance": { "enabled": true, "schedule": "at any time" }
}
```

My only remaining question would be why isn't this the default? Perhaps I missed some documentation that talks about why this has to be opted into?

**Reply by rarkins** [reply]
Although nobody is forced to install Renovate at gun point, there's a vocal minority who complain loudly about too many PRs. Lock file maintenance is a feature we think is a good idea but disable by default for that reason

**Reply by ryancausey** [reply]
Thanks for the clarification. I certainly like renovate and am glad it exists. Thanks to everyone for keeping it going!

**Comment by ryancausey** [comment, marked as answer]
As of version 27.8.0 the provider supports [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy) `update-lockfile` which solves the issue I was having.

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "rangeStrategy": "update-lockfile"
}
```

The old answer for pre-27.8.0 is below.

For future travelers: the solution was to use [lockfileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) as noted below:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "lockFileMaintenance": { "enabled": true, "schedule": "at any time" }
}
```

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/releases/tag/27.8.0

`rangeStrategy=lockfile` is now supported too.

**Reply by rarkins** [reply]
`update-lockfile`?

**Reply by ryancausey** [reply]
How am I able to determine what the behavior of "auto" is for the Terraform manager?

**Reply by rarkins** [reply]
`auto` defaults to `replace` unless specified otherwise.
